test/distributed/checkpoint/e2e/test_e2e_save_and_load.py

```diff
@@ -0,0 +1,135 @@
+from torch.distributed.fsdp.api import ShardingStrategy
+from torch.distributed.tensor.parallel import PairwiseParallel, parallelize_module
+from torch.testing._internal.common_utils import (
+    instantiate_parametrized_tests,
+    parametrize,
+    run_tests,
+)
+
+from torch.testing._internal.distributed._tensor.common_dtensor import (
+    DTensorTestBase,
+    skip_if_lt_x_gpu,
+    with_comms,
+)
+from torch.testing._internal.distributed.checkpoint_utils import with_temp_dir
+
+# Simple and boring model to test interface and some corner cases that do not
+# require complicated wrapping strategy.
+class TestDummyModel(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.net1 = nn.Sequential(nn.Linear(8, 16), nn.ReLU())
+        self.net2 = nn.Sequential(nn.Linear(16, 32), nn.ReLU())
+        self.net3 = nn.Linear(32, 64)
+        self.net4 = nn.Sequential(nn.ReLU(), nn.Linear(64, 8))
+
+    def forward(self, x):
+        return self.net4(self.net3(self.net2(self.net1(x))))
+
+    def get_input(self):
+        return torch.rand(8, 8, device="cuda")
+
+
+class ModelType(Enum):
+    FSDP = auto()
+    HSDP = auto()
+    FSDP_TP = auto()
+
+
+class TestE2ELoadAndSave(DTensorTestBase):
+    def _create_model(self, compile, model_type):
+        dummy_model = TestDummyModel().cuda()
+
+        assert model_type in ModelType, f"{model_type} is not supported."
+        if model_type == ModelType.FSDP:
+            device_mesh = init_device_mesh(self.device_type, (self.world_size,))
+            model = FSDP(
+                dummy_model,
+                device_mesh=device_mesh,
+                use_orig_params=True,
+            )
+        elif model_type == ModelType.HSDP:
+            device_mesh = init_device_mesh(self.device_type, (2, self.world_size // 2))
+            model = FSDP(
+                dummy_model,
+                device_mesh=device_mesh,
+                use_orig_params=True,
+                sharding_strategy=ShardingStrategy.HYBRID_SHARD,
+            )
+        elif model_type == ModelType.FSDP_TP:
+            mesh_2d = init_device_mesh(
+                self.device_type, (2, self.world_size // 2), mesh_dim_names=("dp", "tp")
+            )
+            tp_mesh = mesh_2d["tp"]
+            dp_mesh = mesh_2d["dp"]
+            model = parallelize_module(dummy_model, tp_mesh, PairwiseParallel())
+            model = FSDP(model, device_mesh=dp_mesh, use_orig_params=True)
+
+        optim = torch.optim.Adam(model.parameters(), lr=0.1)
+
+        if compile:
+            model = torch.compile(model)
+
+        model(model.get_input()).sum().backward()
+        optim.step()
+
+        return model, optim
+
+    def _equal_state_dict(self, model_0, model_1):
+        for params_0, params_1 in zip(model_0.values(), model_1.values()):
+            if not torch.equal(params_0, params_1):
+                return False
+        return True
+
+    @with_comms
+    @skip_if_lt_x_gpu(2)
```

Comment:
I don't think FSDP + TP work for GPU = 2. Your local test is probably fine, since we have 8 GPUs on DEVGPU. We will need @skip_if_lt_x_gpu(4).